### PR TITLE
[ASR Refactor][A-PR2] ARK-ASR: pre-lm encoder

### PR DESCRIPTION
**Pull request by AkazaAkane**
## Motivation

ARK-ASR currently computes audio features during multimodal embedding construction on the LM path. Moving that work into a pre-LM encoder prevents audio encoding from stalling the running decode batch.

The first version of this PR performed the encode from `request_builder()`, but the builder blocked until GPU completion. That coupled encoder batching to the number of request-build workers and required raising the worker count from 2 to 8. This rewrite separates request construction from encoder completion:

```text
request builder -> submit encoder work -> pending admission
                                           |
                                           v
                              encoder future completes
                                           |
                                           v
                                    waiting_queue -> LM
```

Only the scheduler thread admits completed requests to `waiting_queue`.

## Modifications

- Add a generic `DeferredAdmission` result for request builders whose output has an asynchronous dependency.
- Add a scheduler-owned pending-admission stage with lifecycle handling for:
  - completed and failed dependencies;
  - request aborts without cancelling shared futures;
  - active-request accounting, idle polling, admin model info, and shutdown.
- Add a bounded pre-LM encoder queue. Existing users retain the unbounded default; ARK-ASR uses a pending limit of 32.
- Add nonblocking `ArkasrPreLMEncoderService.submit_item()` while retaining `encode_item()` as a synchronous compatibility helper.
- Preserve ARK-ASR's existing behavior:
  - CPU-backed embedding cache;
  - single-flight deduplication, now with a completion future per request;
  - private CUDA stream and batch synchronization;
  - greedy batches of up to 8 and model-internal encoder microbatches of up to 8;
  - existing validation, failure recovery, and OOM cleanup.
- Restore request construction defaults to 2 workers and 16 pending builds. Encoder backpressure is now controlled independently through `pre_lm_max_pending=32`.
- Add telemetry for submitted/pending requests, queue saturation, maximum queue depth, batching, queue wait, cache behavior, and encoder time.
- Add focused coverage for deferred admission, abort/error handling, asynchronous submission, asynchronous single-flight, full batch formation, bounded queue backpressure, configuration, and lifecycle wiring.
- Update the ARK-ASR cookbook documentation.

## Related Issues

- Builds on the batched ARK-ASR encoder work in #1411.

## Accuracy Test

Not run in this environment. The model computation, cache format, CUDA stream, synchronization point, and encoder batch-size bound are unchanged by this rewrite. WER parity still needs GPU validation before merge.

## Benchmark & Profiling

| Cache | Concurrency | Main req/s | PR req/s | Throughput Δ | Main mean latency | PR mean latency | Latency Δ |
|---|---|---|---|---|---|---|---|
| Cold | c16 | 32.938 | 55.258 | +67.8% | 0.485 s | 0.287 s | −40.8% |
| Warm | c16 | 98.963 | 98.318 | −0.7% | 0.161 s | 0.162 s | +0.6% |
| Cold | c32 | 48.839 | 100.055 | +104.9% | 0.652 s | 0.317 s | −51.4% |
| Warm | c32 | 136.021 ± 1.377 | 134.717 ± 4.656 | −1.0% | 0.233 ± 0.003 s | 0.236 ± 0.008 s | +1.1% |

## Validation

- Focused pre-commit checks passed across all 17 changed files.
- `git diff --check origin/main...HEAD` passed.
- Rebase `range-diff` confirmed the original two PR patches remained equivalent before the four rewrite commits.
- Unit tests and GPU benchmarks were not executed, per repository task instructions.

## Checklist

- [x] Format code with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the `run-ci` label. Once labeled, every subsequent push re-triggers CI as long as the label remains. Use `/tag-and-rerun-ci higgs` or `/tag-and-rerun-ci moss` to select a TTS CI model, and `/tag-and-rerun-ci fun-asr` or `/tag-and-rerun-ci qwen3-asr` to select an ASR CI model. One selector from each family can be combined, for example `/tag-and-rerun-ci moss fun-asr`. Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/models/arkasr/config.py [hunk 1]
```diff
@@ -29,6 +29,14 @@ class ArkasrPipelineConfig(PipelineConfig):
                 "max_new_tokens": 256,
                 "request_build_max_workers": 2,
                 "request_build_max_pending": 16,
+                "enable_pre_lm_encoder": True,
+                "pre_lm_cache_max_entries": 4096,
+                "pre_lm_cache_size_bytes": 2 * 1024**3,
+                # one drained group maps to exactly one encoder microbatch at
```
Comment (SandyLuXY) [inline sglang_omni/models/arkasr/config.py]: Please follow the comment style `# note (xxx): `

File: sglang_omni/models/arkasr/engine_builder.py [hunk 2]
```diff
@@ -89,9 +120,28 @@ def setup_model_resources(
         *,
         generation_cuda_graph_enabled: bool,
     ) -> None:
-        del server_args, generation_cuda_graph_enabled
+        del generation_cuda_graph_enabled
         model.set_encoder_max_batch_size(self.encoder_max_batch_size)
         init_mm_embedding_cache(self.mm_embedding_cache_size_bytes)
+        if self.enable_pre_lm_encoder:
+            # constructed after SGLang's generation CUDA graphs so the
```
Comment (SandyLuXY) [inline sglang_omni/models/arkasr/engine_builder.py]: same here

File: sglang_omni/models/arkasr/request_builders.py [hunk 3]
```diff
@@ -153,7 +160,13 @@ def request_builder(payload: StagePayload) -> ArkASRRequestData:
             modality=Modality.AUDIO,
             hash=prepared.fingerprint_int,
             feature=features,
-            model_specific_data={"feature_attention_mask": feature_attention_mask},
+            model_specific_data={
+                "feature_attention_mask": feature_attention_mask,
+                # the pre-LM encoder service reads these to split batched
```
Comment (SandyLuXY) [inline sglang_omni/models/arkasr/request_builders.py]: same here

File: tests/unit_test/arkasr/test_pipeline.py [hunk 4]
```diff
@@ -175,22 +275,66 @@ def _fake_defer_infra(server_args, gpu_id, **kwargs):
...
+    stub = _stub_arkasr_engine_build(
+        monkeypatch, want_cuda_graph=False, encoder_service=encoder_service
+    )
+
+    scheduler = create_sglang_arkasr_executor("AutoArk-AI/ARK-ASR-3B")
+
+    assert stub.adapter_kwargs["audio_encoder_service"] is encoder_service
+    assert scheduler.shutdown_callback == encoder_service.close
+    # A-PR4's model-internal bound is still applied when the pre-LM service is
+    # in front of it; the two batch limits are independent.
```
Comment (SandyLuXY) [inline tests/unit_test/arkasr/test_pipeline.py]: Please don't leave `A-PR4` in the comment, it's only local for the tracker

File: tests/unit_test/arkasr/test_encoder_service.py
Comment (SandyLuXY) [inline tests/unit_test/arkasr/test_encoder_service.py:1]: Please help update `tests/README.md`

File: sglang_omni/scheduling/omni_scheduler.py [hunk 5]
```diff
@@ -1567,15 +1655,21 @@ def start(self) -> None:
...
             finally:
+                self._discard_pending_request_admissions()
                 self._shutdown_resources()
 
     def event_loop(self) -> None:
         self.start()
 
     def stop(self) -> None:
         self._running = False
+        self._discard_pending_request_admissions()
```
Comment (SandyLuXY) [inline sglang_omni/scheduling/omni_scheduler.py:1695]: seems this will make three existing lifecycle tests fail because their `object.__new__` schedulers do not initialize `_request_admission_lock` or `_pending_request_admissions`

**Review by SandyLuXY — CHANGES_REQUESTED** [review]

File: sglang_omni/scheduling/omni_scheduler.py (hunk 5 above)
Comment (AkazaAkane) [reply sglang_omni/scheduling/omni_scheduler.py:1695]: Added _request_admission_lock = threading.RLock().
Added _pending_request_admissions = {}.
Updated affected fixtures in:tests/unit_test/pipeline/test_scheduler.py
tests/unit_test/pipeline/test_stage.py

File: tests/unit_test/arkasr/test_encoder_service.py
Comment (AkazaAkane) [reply tests/unit_test/arkasr/test_encoder_service.py:1]: Updated.

File: tests/unit_test/arkasr/test_pipeline.py (hunk 4 above)
Comment (AkazaAkane) [reply tests/unit_test/arkasr/test_pipeline.py]: Deleted

File: sglang_omni/models/arkasr/request_builders.py (hunk 3 above)
Comment (AkazaAkane) [reply sglang_omni/models/arkasr/request_builders.py]: Fixed.

File: sglang_omni/models/arkasr/engine_builder.py (hunk 2 above)
Comment (AkazaAkane) [reply sglang_omni/models/arkasr/engine_builder.py]: Fixed.

File: sglang_omni/models/arkasr/config.py (hunk 1 above)
Comment (AkazaAkane) [reply sglang_omni/models/arkasr/config.py]: Fixed.

File: tests/unit_test/scheduling/test_deferred_admission.py
Comment (SandyLuXY) [inline tests/unit_test/scheduling/test_deferred_admission.py:1]: this file is still missing from test readme

File: tests/unit_test/arkasr/test_encoder_service.py [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+    futures = [service.submit_item(item) for item in items]
+    deadline = time.monotonic() + 2
+    while model.encode_calls == 0 and time.monotonic() < deadline:
+        time.sleep(0.005)
+    assert model.encode_calls == 1
+    gate.set()
+    for future in futures:
+        future.result(timeout=2)
+
+    assert model.encode_batch_sizes == [1, 8]
```
Comment (SandyLuXY) [inline tests/unit_test/arkasr/test_encoder_service.py:174]: There is no synchronization ensuring the worker consumes item 1 before the other eight submissions, so the valid greedy schedule `[8, 1]` fails this assertion on the current head.

**Review by SandyLuXY — CHANGES_REQUESTED** [review]

File: tests/unit_test/scheduling/test_deferred_admission.py
Comment (AkazaAkane) [reply tests/unit_test/scheduling/test_deferred_admission.py:1]: added

File: tests/unit_test/arkasr/test_encoder_service.py (hunk 6 above)
Comment (AkazaAkane) [reply tests/unit_test/arkasr/test_encoder_service.py:174]: The test now submits item 1, waits until it is actively encoding, queues the remaining eight, then releases the gate—making [1, 8] deterministic.

File: tests/unit_test/scheduling/test_deferred_admission.py
Comment (AkazaAkane) [reply tests/unit_test/scheduling/test_deferred_admission.py:1]: image

**Review by SandyLuXY — APPROVED** [review]
lgtm
